File: qc-core/src/main/java/com/eager/questioncloud/core/domain/payment/infrastructure/QuestionPaymentHistoryDocumentReadConverter.java
package com.eager.questioncloud.core.domain.payment.infrastructure;

import com.eager.questioncloud.core.domain.coupon.enums.CouponType;
import com.eager.questioncloud.core.domain.payment.enums.QuestionPaymentStatus;
import com.eager.questioncloud.core.domain.payment.model.QuestionPaymentCoupon;
import com.eager.questioncloud.core.domain.payment.model.QuestionPaymentHistory.QuestionPaymentHistoryOrder;
import com.eager.questioncloud.core.domain.question.enums.Subject;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Date;
import java.util.List;
import java.util.Optional;
import org.bson.Document;
import org.springframework.core.convert.converter.Converter;

public class QuestionPaymentHistoryDocumentReadConverter implements Converter<Document, QuestionPaymentHistoryDocument> {
    @Override
    public QuestionPaymentHistoryDocument convert(Document source) {
        return QuestionPaymentHistoryDocument.builder()
            .paymentId(source.getLong("_id"))
            .orderId(source.getString("orderId"))
            .userId(source.getLong("userId"))
            .orders(convertOrders(source.getList("orders", Document.class)))
            .coupon(convertCoupon(source))
            .amount(source.getInteger("amount"))
            .isUsedCoupon(source.getBoolean("isUsedCoupon"))
            .status(QuestionPaymentStatus.valueOf(source.getString("status")))
            .createdAt(convertDate(source.get("createdAt", Date.class)))
            .build();
    }

    private List<QuestionPaymentHistoryOrder> convertOrders(List<Document> orders) {
        return orders.stream()
            .map(orderDocument -> new QuestionPaymentHistoryOrder(
                orderDocument.getLong("questionId"),
                orderDocument.getInteger("amount"),
                orderDocument.getString("title"),
                orderDocument.getString("thumbnail"),
                orderDocument.getString("creatorName"),
                Subject.valueOf(orderDocument.getString("subject")),
                orderDocument.getString("mainCategory"),
                orderDocument.getString("subCategory")
            ))
            .toList();
    }

    private QuestionPaymentCoupon convertCoupon(Document source) {
        if (!source.getBoolean("isUsedCoupon", false)) {
            return null;
        }

        Document coupon = source.get("coupon", Document.class);
        if (coupon == null) {
            return null;
        }

        return new QuestionPaymentCoupon(
            getEmbeddedValue(coupon, "userCouponId", Long.class),
            getEmbeddedValue(coupon, "title", String.class),
            Optional.ofNullable(getEmbeddedValue(coupon, "couponType", String.class))
                .map(CouponType::valueOf)
                .orElse(null),
            getEmbeddedValue(coupon, "value", Integer.class)
        );
    }

    private <T> T getEmbeddedValue(Document document, String key, Class<T> clazz) {
        return document.get(key, clazz);
    }

    private LocalDateTime convertDate(Date date) {
        return date != null
            ? date.toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime()
            : null;
    }
}

File: qc-core/src/main/java/com/eager/questioncloud/core/domain/payment/infrastructure/document/QuestionPaymentHistoryDocumentReadConverter.kt
package com.eager.questioncloud.core.domain.payment.infrastructure.document

import com.eager.questioncloud.core.domain.coupon.enums.CouponType
import com.eager.questioncloud.core.domain.payment.enums.QuestionPaymentStatus
import com.eager.questioncloud.core.domain.payment.model.QuestionPaymentCoupon
import com.eager.questioncloud.core.domain.payment.model.QuestionPaymentHistoryOrder
import com.eager.questioncloud.core.domain.question.enums.Subject
import org.bson.Document
import org.springframework.core.convert.converter.Converter
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.*

class QuestionPaymentHistoryDocumentReadConverter :
    Converter<Document, QuestionPaymentHistoryDocument> {
    override fun convert(source: Document): QuestionPaymentHistoryDocument {
        return QuestionPaymentHistoryDocument(
            orderId = (source.getString("_id")),
            userId = (source.getLong("userId")),
            orders = (convertOrders(source.getList("orders", Document::class.java))),
            coupon = (convertCoupon(source)),
            amount = (source.getInteger("amount")),
            isUsedCoupon = (source.getBoolean("isUsedCoupon")),
            status = (QuestionPaymentStatus.valueOf(source.getString("status"))),
            createdAt = (convertDate(source.get("createdAt", Date::class.java)))
        )
    }

    private fun convertOrders(orders: List<Document>): List<QuestionPaymentHistoryOrder> {
        return orders.stream()
            .map { orderDocument: Document ->
                QuestionPaymentHistoryOrder(
                    orderDocument.getLong("questionId"),
                    orderDocument.getInteger("amount"),
                    orderDocument.getString("title"),
                    orderDocument.getString("thumbnail"),
                    orderDocument.getString("creatorName"),
                    Subject.valueOf(orderDocument.getString("subject")),
                    orderDocument.getString("mainCategory"),
                    orderDocument.getString("subCategory")
                )
            }
            .toList()
    }

    private fun convertCoupon(source: Document): QuestionPaymentCoupon? {
        if (!source.getBoolean("isUsedCoupon", false)) {
            return null
        }

        val coupon = source.get("coupon", Document::class.java)

        return QuestionPaymentCoupon(
            coupon.getLong("userCouponId"),
            coupon.getString("title"),
            CouponType.valueOf(coupon.getString("couponType")),
            coupon.getInteger("value")
        )
    }

    private fun convertDate(date: Date): LocalDateTime {
        return date.toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime()
    }
}